source/common/buffer/watermark_buffer.cc

```diff
@@ -72,7 +72,9 @@ void WatermarkBuffer::setWatermarks(uint32_t low_watermark, uint32_t high_waterm
   ASSERT(low_watermark < high_watermark || (high_watermark == 0 && low_watermark == 0));
   low_watermark_ = low_watermark;
   high_watermark_ = high_watermark;
-  checkHighWatermark();
+  overflow_watermark_ = high_watermark * runtime_.threadsafeSnapshot()->getInteger(
```

Comment:
From a perf perspective, I would probably snap the multiplier at buffer creation time vs. reading it on every operation.